### DISCUSSION TITLE
SavvyCAN and Arduino/MCP2515

### DISCUSSION THREAD

**Discussion by lares67** [collin80/SavvyCAN#448] [Q&A]
Hi colin80!
I build a Arduino MCP2515 CAN interface like this one:
https://github.com/autowp/arduino-mcp2515

It is working with a "firmware" like this:
https://github.com/autowp/can-usb

I like your SavvyCAN analyzer a lot! 
I use it at a moment offline - I read the can traffic and save it to a file, then I load it to SavvyCAN as a GVRET csv file and try to analyze it.

Are there any chances, that SavvyCAN will at some point be able to communicate with this type of interfaces?

You know - offline is good - online is better :)

Best regards
Mat

PS. Any chance to use SavvyCAN with my old Windows XP laptop, that I use to work in my car (I am not skilled enough to compile it myself...)?

**Comment by collin80** [comment]
Well, here's the thing, LAWICEL(which is what the linked firmware is) devices should work with SavvyCAN as there is a LAWICEL to socketcan interface in LINUX. However, if you aren't using LINUX then that does you no good. I suppose it could be possible to directly support LAWICEL devices in SavvyCAN but I have not done any work to support it. 

As for Windows XP support - probably not. The current codebase of SavvyCAN targets QT 5.15 which supports only back to Windows 7. Also, all the recent builds of SavvyCAN are 64 bit. It can be compiled 32 bit but I haven't bothered in a while. It's possible that a 32 bit compiled version of SC might run on Windows XP but no promises there.

**Reply by ghost** [reply]
What firmware should I use to be able to run arduino nano + mcp2515 with your app? I am using Ubuntu 22.04. Thanks in advance!

**Reply by collin80** [reply]
Well, I don't have any firmware for an arduino nano. So, nothing written by me will be of any use. I thought that people had done this themselves though. I wish I could remember anyone who has written the firmware you need. I suppose I really ought to get lawicel support into SavvyCAN to make it easier for devices like yours to work. It's much more likely for you to find a lawicel compatible firmware than gvret compatible.

**Reply by ghost** [reply]
OK. In short, I ordered myself such a thing.  https://youtu.be/62txKc7EtCA?t=113 I think it should work. Thanks for the answer!

**Comment by lares67** [comment]
Thank you for your answer!

I regret it more and more - I'm an absolute zero, if I get to a linux machine...

But if linux has an inbuilt interface for Lawicel, I will give it a try - I have an experimental installation of linux on one of my old computers but... it will be a real pain ;) I'v never really worked with linux...

Now I partially can understand, why you did not bother to write a native support for Lawicel in SavvyCAN. 

Would it be a complicated task? It would make my life so much easier ;)

As I discovered the "Range State Window" in SavvyCan, it remembered me at some paper I read before - here is a link to it:
https://web.eecs.umich.edu/~mpese/papers/ACM_CCS_LibreCAN_Paper.pdf

I consider it as highly interesting - maybe you could find some ideas there in interesting too.

Best regards
Mat

**Comment by consp** [comment]
I wrote an slcan compatible version which might/should be lawicel compatible but I have no way of testing it. If you can compile SavvyCAN yourself you might want to try that. If it works (or you give input to what goes wrong) I might be able to get it somewhat ready to make a pull request.

This does not use any QTSerial/CAN translations or socketcan/slcan and should be easier to use on non linux systems. [See the slcan-connection-method here.](https://github.com/consp/SavvyCAN/tree/slcan-connection-method)

Note: not tested on Windows or OSX since I have no setup to compile savvycan on those platforms but due to it using all standard serial interface parts it should be compatible with those OS's.

**Comment by lares67** [comment]
I'm afraid, I'm not that skilled to make it happen. I'm absolut novice at Linux...

Till now, I only used compiled versions of SavvyCan and was not really successful in getting it to work on my Linux machine using its LAWICEL to socketcan interface...

And my Linux machine is dying (video card is malfunctioning recently). The only other machine I have, I could use for testing, is an old Win XP machine (which will not work with Win7, I suppose), and this is not enough to compile SavvyCan (?).

Despite of this all, I will try to read some stuff, try to compile your version, but it will take time and a success is pretty unlikely ;)

On the "old" SavvyCan forum on www.savvycan.com I found two other questions to lawicel support:
https://www.savvycan.com/index.php/forum/suggestion-box/190-candapter-support
https://www.savvycan.com/index.php/forum/welcome-mat/208-new-can-usb-device-support-which-way-do-you-recommend

Maybe you can contact this people too, and maybe they are more skilled, then I am ;)

Thank you for your answer and your programming effort!

I will definitely follow your fork on github!

Mat

**Reply by consp** [reply]
I'll see if I can make a Windows debug build, will take a moment since I do not have any proper windows installation available right now. I kind of need someone to test it due to lacking an actual proper slcan/lawicel device so you probably just volunteered for that. I just have my device with lawicel support which should be compatible but it might not be completely compatible. looks like mingw is included qt for windows so it should not be too much of an issue to build it.

**Reply by lares67** [reply]
I'll try to do my best ;)

In the first post in this thread you find the interface, I have and the firmware I use on it.

Mat

PS. You remember - my Windows is the ancient XP 32 bit ;)

**Reply by consp** [reply]
> PS. You remember - my Windows is the ancient XP 32 bit ;)

Missed that...

That's definitely extremely unlikely to happen as Qt 5.6.3 is the latest supporting Windows XP (64bit non the less) and SavvyCAN requires at least Qt 5.14 (Preferably 5.15) to have proper serial bus support so you are out of luck there. I'm pretty sure you can run windows 7 on most pc/laptops which run XP unless they are truly ancient. I'll try to compile one with 32/64 bit support for 5.15 and you can see if it works or not. I'm pretty sure it will crash on XP though due to missing win32 api's.

https://github.com/consp/SavvyCAN/releases/tag/SLCAN-test-0

**Comment by lares67** [comment]
Just tried te SavvyCAN-i386-Win32-SLCANtest-Debug.zip on the WIN_XP_32.

Error - "No entry point found to procedure CancelIoEx in the library KERNEL332.dll" (translated from polish, so...)

Mat

**Reply by consp** [reply]
I'm not surprised, Kernel32 is missing functions which it needs to run. Won't work then. I Can't compile it for WinXP since it needs a too old version of QT.  It was worth a try, but no luck unfortunately.
